### PR TITLE
[koa-bodyparser] add ability to strictly type request body

### PR DESCRIPTION
**Pull request by pkuczynski**
I am trying to implement the ability to strictly type the request body. Currently, `body` is defined as `any` so any attempt to override the `ctx: { request: { body } }` with a more specific type, will anyway result in all the props being typed as `any` due to the way `koa` defines context as the intersection type.

I created two separate test suites. One for a strictly typed body and one without it. After changing `body: unknown`, it works perfectly for my strictly typed tests but fails the untyped one.

```js
ctx.request.body.a
// TS2571: Object is of type 'unknown'.
```

@anup-2s, @hirochachacha, @amaschas @DavidCai1993 @jkeylu @brikou @harryparkdotio @chatoo2412 @tellnes do you have any idea how this could be achieved better?

### CODE REVIEW

File: types/koa-bodyparser/index.d.ts [hunk 1]
```diff
@@ -21,7 +22,7 @@ import * as Koa from "koa";
 
 declare module "koa" {
     interface Request {
-        body?: any;
+        body: unknown;
```
Comment (harryparkdotio) [inline types/koa-bodyparser/index.d.ts:25]: this should remain as an optional field

**Comment by harryparkdotio** [comment]
Currently, the approach I use in my koa applications to type the request body is to accept the body as an `any` type, validate it with a validation library (I often use joi), and from their, I then type the request body given that I can guarantee that the request body matches the type post validation. This is because there is no way to force the type of the request body to be that of the actual request, typescript does not support runtime types, or casting to a type like in ASP.NET for example.

Due to this, it is safest to actually use the any type for the incoming request body, since there is nothing requiring the types to match, and in actuality, it provides a false sense of security.

Interested to hear your thoughts on this, but I am of the opinion that this can, and probably should be achieved through validation, or if you want to work around this, using type casting is a sufficient alternative.

File: types/koa-bodyparser/index.d.ts (hunk 1 above)
Comment (pkuczynski) [reply types/koa-bodyparser/index.d.ts:25]: Actually according to the `koa-bodyparser` if body is empty, it's still an empty hash:

> // the parsed body will store in ctx.request.body
>  // if nothing was parsed, body will be an empty object {}

**Comment by pkuczynski** [comment]
@harryparkdotio I do the same, although I am using `ajv` in a middleware. This works very well (from version 7 ajv plays very well with TS allowing to ensure schema is matching the interface). So I am safe when it comes to data in `request.body`. I just wanna type it, so my middleware is more safe regarding field names, types, etc. Like in this example: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/52025/files#diff-1c7ca2cd2fbd3fe9699b0790b8ca3db453e91205d282faaf85efbe9d83ea7729

Problem is, that when `any` is used, this is not possible, as union of `any & CustomType` will be `any`... :(

And if we will use `unknown`, which could be a more appropriate type in this case, then we will screw all the people, who did not type it.

Maybe someone else will have some other idea? :(

**Comment by harryparkdotio** [comment]
You could override the type in the within your codebase with typescript type declarations, similar to how koa-bodyparser does so for the koa context types.

**Comment by pkuczynski** [comment]
Unfortunately, I can not. If it's defined as `any` it will stay `any`. Have a look at the example tests I provided.

**Comment by pkuczynski** [comment]
@anup-2s, @hirochachacha, @amaschas @DavidCai1993 @jkeylu @brikou @chatoo2412 @tellnes do you have any idea how this could be achieved better?

**Comment by pkuczynski** [comment]
> You could override the type in the within your codebase with typescript type declarations, similar to how koa-bodyparser does so for the koa context types.

Unfortunately when I do that, I am getting an error:

```
TS2687: All declarations of 'body' must have identical modifiers.
TS2717: Subsequent property declarations must have the same type. Property 'body' must be of type 'any', but here has type 'unknown'. 
```

**Comment by pkuczynski** [comment]
@harryparkdotio I think I found the solution. Could you please approve?

**Pull request closed without merging**